<h1>Jupiter Notebook</h1>


In [91]:
print("test")

test


#Part 1

#Ishan Phadte 300238878
#Eeshaan Patel 300165634

#Part 2 

#The Knapsack problem involves maximing value for a small weight. A person has a backpack with X capacity. Their goal is to fill their backpack with items with the highest combined price while holding a weight which is equal or less capacity

In [92]:
#Part 3 (Done)

import pandas as pd
import itertools
import numpy as np

#This Data set was provided by the TAs and hosted in Ishan's Personal Github and is Public 
url = "https://raw.githubusercontent.com/IshanPhadte776/CSI4106Assignment1/main/knapsack_5_items.csv"

dataset = pd.read_csv(url)
dataset.columns

dataset.head(10)

def string_to_list(string):

  string_list = string.strip('[]').split()

  float_list = [float(element) for element in string_list]

  return float_list


dataset = dataset.dropna()

dataset.Weights = dataset.Weights.apply(lambda x : string_to_list(x))
dataset.Prices = dataset.Prices.apply(lambda x : string_to_list(x))
dataset['Best picks'] = dataset['Best picks'].apply(lambda x : string_to_list(x))

In [93]:
#Part 4 Generate and Test

#This solution involves generating all solutions and testing all as well. This solution is similar to the "Guess and Check" method and the simple method

#Done
def gen_and_test(data):

    weights = data[0]
    prices = data[1]
    capacity = data[2]
    bestPicks = data[3]
    target = data[4]
    
    #5
    num_items = len(weights)
    
    #List of Zeros as there isn't a best Solution found yet 
    best_solution = [0] * num_items
    #Best Solution so far 
    best_solution_price = 0
    
    #32 Times
    for i in range(2 ** num_items):
        
        current_weight = 0  # Initialize current_weight to 0
        current_price = 0  # Initialize current_price to 0
        
        
        # Use the bin() function to convert to binary and remove the '0b' prefix
        binary_representation = bin(i)[2:]

        # Calculate the number of zeros to prepend to make it 5 characters long
        num_zeros = 5 - len(binary_representation)

        # Create a list with 5 elements, initially filled with 0s
        binary_list = [0] * num_zeros

        # Extend the list with the binary representation converted to integers
        binary_list.extend([int(bit) for bit in binary_representation])  


        # Iterate through the indices of binary_list and weights
        for j in range(num_items):
            # Multiply the corresponding elements and add to current_weight
            current_weight += binary_list[j] * weights[j]


        # Iterate through the indices of binary_list and prices
        for j in range(num_items):
            # Multiply the corresponding elements and add to current_price
            current_price += binary_list[j] * prices[j]
        
        #If the current solution is better than the previous best solution
        if current_weight <= capacity and current_price > best_solution_price:
            #Update the best solution 
            best_solution = binary_list
            best_solution_price = current_price
    
    #Return the Best Solution 
    return best_solution_price, best_solution

solutions = []
#For Every Row 
for _, row in dataset.iterrows():
    #Take the best price
    target = row['Best price']
    #Call the gen_and_test function 
    solution, indexes = gen_and_test(row)
    #print(solution)

    
    solutions.append(1 if target == solution else 0)
    
print(solutions)


/var/folders/tp/f3ych92x75138hq94725my2r0000gn/T/ipykernel_63417/1954860600.py:8: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  weights = data[0]
/var/folders/tp/f3ych92x75138hq94725my2r0000gn/T/ipykernel_63417/1954860600.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  prices = data[1]
/var/folders/tp/f3ych92x75138hq94725my2r0000gn/T/ipykernel_63417/1954860600.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  capacity = data[

[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 

In [94]:
#Part 5 Greedy Algorithm 

#The Greedy Algorithm involves picking the "Best" items from a weight to price stand point. This doesn't always provide the correct solution but a great start 

#Done
def greedy(data):
    
    weights = data[0]
    prices = data[1]
    capacity = data[2]
    bestPicks = data[3]
    target = data[4]
    
    #5
    num_items = len(weights)
    
    #Create a Slice so we can order the items in order 
    item_indices = list(range(num_items))
        

    # Calculate the value-to-weight ratio for each item (Best in-theory items)
    value_per_weight = []

    for i in range(num_items):
        ratio = prices[i] / weights[i]
        value_per_weight.append(ratio)

    # Sort items by the value-to-weight ratio in descending order
    sorted_indices = []

        # Loop through each item (represented by 'i') to find the one with the highest ratio
    for i in range(num_items):
        max_index = 0  # Filler Values
        max_ratio = -99  # Filler Values
        
        # Iterate through all items again to compare their ratios
        for j in range(num_items):
            # Check if item 'j' is not already in the sorted list
            if j not in sorted_indices:
                # Compare the value-to-weight ratio of item 'j' with the current maximum
                if value_per_weight[j] > max_ratio:
                    # If the ratio of 'j' is higher, update the maximum index and ratio
                    max_index = j
                    max_ratio = value_per_weight[j]

        # Add the index of the item with the highest ratio to the sorted list
        sorted_indices.append(max_index)
        
    #Indices will be sorted based on weight to price ratio 
        

    current_weight = 0
    current_price = 0
    selected_items = [0] * num_items

    #From best to worst value....
    for i in sorted_indices:
        #If the new item can fit in 
        if current_weight + weights[i] <= capacity:
            # Include the entire item if it fits
            #Include the items in order from most bang for buck to least bang for buck 
            selected_items[i] = 1
            #update our variables 
            current_weight += weights[i]
            current_price += prices[i]

    return current_price, selected_items

solutions_greedy = []
for _, row in dataset.iterrows():
    target = row['Best price']
    solution, indexes = greedy(row)
    solutions_greedy.append(1 if target == solution else 0)
        
print("Greedy Accuracy is", np.mean(solutions_greedy))

/var/folders/tp/f3ych92x75138hq94725my2r0000gn/T/ipykernel_63417/975615045.py:8: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  weights = data[0]
/var/folders/tp/f3ych92x75138hq94725my2r0000gn/T/ipykernel_63417/975615045.py:9: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  prices = data[1]
/var/folders/tp/f3ych92x75138hq94725my2r0000gn/T/ipykernel_63417/975615045.py:10: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  capacity = data[2]


Greedy Accuracy is 0.8341937443061038


In [95]:
#Part 6 Simulated Annealing
#This code immplements the simulated annealing algorithm to solves the knapsack problem.
#It calculates the accuracy of the algoritms solutions compared to the prices in the dataset.

#https://uottawa.brightspace.com/d2l/le/content/392261/viewContent/5543376/View 

import random
import math

def simulated_annealing(data, N, initial_temperature, cooling_rate):
    # basic required info
    price = data['Prices']
    weights = data['Weights']
    capacity = data['Capacity']

    n = len(price)
    current_solution = [random.randint(0, 1) for _ in range(n)] #generate the initial candidate solution
    current_knapsack = knapsack(current_solution, price, weights, capacity)#knapsack of current solution
    best_solution = current_solution.copy() #tracker
    best_price = current_knapsack #tracker

    #https://www.youtube.com/watch?v=XNMGq5Jjs5w
    for _ in range(N):
        temp = initial_temperature

        # This solution only looks at one generated neighbor solution, another option to look at multiple neighboring solutions.
        # My personal implementation that looked at multiple solutions decreases the accuracy when having 10 iterations however increases the accuracy at 100.
        # Conclusions to be made, either my implementation is incorrect or looking at multiple neighbors is more accurate when the iterations are higher, and vice versa.
        # Make multiple attempts to improve the solution in each iteration
        for _ in range(n):
            neighbor_solution = current_solution.copy()
            index = random.randint(0, n-1)
            neighbor_solution[index] = 1 - neighbor_solution[index] #knapsack of neighbour to improve solution

            neighbor_knapsack = knapsack(neighbor_solution, price, weights, capacity)
            difference = neighbor_knapsack - current_knapsack 

            if neighbor_knapsack >= current_knapsack or random.random() < math.exp(difference / temp):
                current_solution = neighbor_solution
                current_knapsack = neighbor_knapsack

                #solution is updated if a better solution is found
                if current_knapsack > best_price:
                    best_solution = current_solution.copy()
                    best_price = current_knapsack

        # Adjust temperature using cooling rate
        temp *= cooling_rate

    return best_price, best_solution

#calculates total price or fitness
def knapsack(solution, price, weights, capacity):
    total_price = 0
    total_weight = 0

    for i in range(len(solution)):
        if solution[i] == 1:
            total_price += price[i]
            total_weight += weights[i]

    if total_weight > capacity:
        return 0  #non-satisfiable solution

    return total_price

solutions_sa = []

for _, row in dataset.iterrows():
    target = row['Best price']
    solution, indexes = simulated_annealing(row, N = 10, initial_temperature=1, cooling_rate=0.01)
    solutions_sa.append(1 if target == solution else 0)
    
print("Simulated Annealing Accuracy is", np.mean(solutions_sa))

Simulated Annealing Accuracy is 0.49175017714343555


In [96]:
#Part 7 Genetic Algorithm 
# This code implements the genetic algorithm to find the best solution that maximizes the total price.
# It repeatedly evolves the population for 50 generations using the crossover and mutation functions to find its answer.

import random
import numpy as np
import pandas as pd

#calculate total price or fitness
def calculate_fitness(ind, prices, weights, capacity):
    fitness = sum(ind[i] * prices[i] for i in range(len(ind)))
    total_weight = sum(ind[i] * weights[i] for i in range(len(ind)))
    
    if total_weight > capacity:
        return 0  #non-satisfiable solution
    else:
        return fitness

#required function to maintain genetic diversity and not have copies of the parents
def crossover(parent1, parent2, cross_rate):
    #performs crossover between two parents 
    if random.random() < cross_rate:
        crossover_point = random.randint(1, len(parent1) - 1)
        #recombine the parents in opposite ways
        child1 = parent1[:crossover_point] + parent2[crossover_point:]
        child2 = parent2[:crossover_point] + parent1[crossover_point:]
    else:
        child1, child2 = parent1.copy(), parent2.copy()
    
    return child1, child2

def mutation(child, mut_rate):
    for i in range(len(child)):
        if random.random() < mut_rate:
            child[i] = 1 - child[i]  # Flip if random is less then mut_rate
    return child

def genetic_algorithm(data, population_size, num_generations, mut_rate, cross_rate, tournament_size):
    # basic required info
    prices = np.fromstring(data['Prices'][1:-1], dtype=int, sep=' ')
    weights = np.fromstring(data['Weights'][1:-1], dtype=int, sep=' ')
    capacity = data['Capacity']
    
    # creates a start population with random solutions
    population = [random.choices([0, 1], k=len(prices)) for _ in range(population_size)]
    
    best_solution = None
    best_solution_price = 0
    
    for generation in range(num_generations):
        #calulate the fitness of each individual in the population
        fitness_scores = [calculate_fitness(ind, prices, weights, capacity) for ind in population]
        
        #select parents for the next generation using a tournament 
        parents = []
        for _ in range(population_size):
            tournament = random.sample(range(population_size), tournament_size)
            winner = max(tournament, key=lambda x: fitness_scores[x])
            #best fitness is parent
            parents.append(population[winner])
        
        #create the next generation by crossover and mutation
        next_generation = []
        for i in range(0, population_size, 2):
            if i + 1 < population_size:  #check for a pair of parents
                parent1, parent2 = parents[i], parents[i + 1]
                child1, child2 = crossover(parent1, parent2, cross_rate)#crossover func used - refer to code above
                child1 = mutation(child1, mut_rate) #mutation func used for line below too- refer to code above
                child2 = mutation(child2, mut_rate)
                next_generation.extend([child1, child2])
            else:
                #odd number of parents, copy the last one to the next generation
                next_generation.append(parents[i])

        #update
        population = next_generation
        
        #find the best solution
        generation_best_price = max(fitness_scores)
        if generation_best_price > best_solution_price:
            best_solution = population[fitness_scores.index(generation_best_price)]
            best_solution_price = generation_best_price
    
    return best_solution_price, best_solution

url = "https://raw.githubusercontent.com/IshanPhadte776/CSI4106Assignment1/main/knapsack_5_items.csv"
dataset = pd.read_csv(url)

#please be patient as the code takes a bit of time to execute a pop_size of 50
solutions_ga = []
for _, row in dataset.iterrows():
    target = row['Best price']
    solution, indexes = genetic_algorithm(row, population_size=50, num_generations=50, mut_rate=0.1, cross_rate=0.7, tournament_size=5)
    solutions_ga.append(1 if target == solution else 0)
    
print("Genetic Algorithm Accuracy is", np.mean(solutions_ga))

Genetic Algorithm Accuracy is 0.9996


#Part 8 Comparative Study (Eeshaan) 

Using the numbers provided in the template we can make the assumption that the genetic algoritm provides
the highest level of accuracy amongst the three algorithms. Greedy provides an accuracy of roughly 83%,
simulated annealing provides an accuracy of roughly 50%, and genetic provides an accuracy of roughly 99%.


The tables below display the affects on accuracy depending on how the paramenters are changed
| Simulated Annealing Algorithm| N | Initial Temperature | Cooling Rate|
| ------------- |:--------------:| --------------:| --------------:|
| If parameters were increased | Increases | Increases | Same |
| If parameters were decreased | Decreases | Decreases | Same |

| Genetic Algorithm| Population Size | Number of Generations | Mutation Rate| Crossover Rate| Tournament Size|
| ------------- |:--------------:| --------------:| --------------:|--------------:| --------------:|
| If parameters were increased | Increases | Increases | Decreases |Slight Increase | Slight Decrease |
| If parameters were decreased | Decreases | Decreases | Decreases |Slight Decrease | Slight Increase |





#Part 9 Conclusion (Ishan)

In Conclusion, The Generative and Test Solution provided a perfect result with 100% accuracy. However, the algorithm was only looking thorough 32 different solutions so the runtime of O(2^n) was not an issue but with larger n values, speed will be an issue

The Greedy Algorithm was effective at roughly 83% accuracy with an O(n^2) speed which was very effective. 

The Simulated Annealing had a poor accuracy sitting at roughly 50%

The Genetic Algorithm was just as effective as the Greedy Algo with 80% accuracy. It also used an O(n^2) speed as well 


Overall, the Greedy and Genetic Algorithm were the best solutions if the scope of the project was expanded. However, if the scope remained at the current size, the generate and test solution is the best. 

#Part 10 References 
Fall2023-CSI4106-PopulationBased - Course slidshow was also referred too.

Generate and Test Help
#https://www.geeksforgeeks.org/generate-and-test-search/

Greedy Algorithm
#https://www.geeksforgeeks.org/greedy-algorithms/

Knapsack Solution
#https://www.geeksforgeeks.org/0-1-knapsack-problem-dp-10/

Simulated Annealing 
#https://www.youtube.com/watch?v=XNMGq5Jjs5w

Genetic 
#https://machinelearningmastery.com/simple-genetic-algorithm-from-scratch-in-python/
